R-package/src/xgboost_R.cc

```diff
@@ -682,6 +682,7 @@ XGB_DLL SEXP XGProxyDMatrixCreate_R() {
 
 XGB_DLL SEXP XGProxyDMatrixSetDataDense_R(SEXP handle, SEXP R_mat) {
   R_API_BEGIN();
+  R_SetExternalPtrProtected(handle, R_mat);
```

Comment:
Hmm, so when is the data released?

Reply:
My understanding is that the external pointer object has only one slot for a protected `SEXP` object. So a call to this function would make the previous object in that slot become unprotected:
https://github.com/wch/r-source/blob/872770ededfc26429adca09fab0803e1b78b6eac/src/main/memory.c#L3855
https://github.com/wch/r-source/blob/872770ededfc26429adca09fab0803e1b78b6eac/src/main/memory.c#L1292

But I'm not 100% sure about it.

Reply:
Going through C macros is not my strong suit ...

Do you think it's difficult to store the data in the `ProxyDMatrixWrapper` you wrote in `xgboost_R.cc`? Currently, it's not used in the R code, but we can make the R proxy dmatrix creator return the wrapper instead of the real proxy dmatrix handler.

Reply:
In Python, I store the temporary data in the iterator, but anywhere is fine.

Reply:
Going through the code again, it looks like both `xgb.ExtMemDMatrix` and `xgb.QuantileDMatrix.from_iterator` will end up calling one of these 3 functions in the end, so please ignore the latter comment.

> Do you think it's difficult to store the data in the ProxyDMatrixWrapper you wrote in xgboost_R.cc?

I think it could be quite tricky, because the public R C API doesn't allow protecting/unprotecting individual objects (unless by the logic of putting an unprotected object inside a protected container like list/externalptr) - it only offers `UNPROTECT(<last N>)`.

It'd have to use a list object and be constantly assigning the same element to it in order to achieve the same effect, which wouldn't have any advantage compared to the mechanism in this PR of assigning to the "protected" slot of the externalptr, **assuming that it works the way we expect it to**.

Reply:
> assuming that it works the way we expect it to.

That's internal to R right? Even if it works today, it can break at the next release and we don't have a way to test that.

Reply:
> > assuming that it works the way we expect it to.
> 
> That's internal to R right? Even if it works today, it can break at the next release and we don't have a way to test that.

No, that's part of the public C API of R, which has pretty strong guarantees. Internal API calls in package code would disqualify packages from being hosted at CRAN.

And either way, it looks like it should end up using the macro from the earlier link that decreases the reference count, unless you manually happen to define internal-use variables before including the R header:
https://github.com/wch/r-source/blob/872770ededfc26429adca09fab0803e1b78b6eac/src/include/Rinternals.h#L191

Reply:
> No, that's part of the public C API of R

hmm, I meant the "external pointer object has only one slot for a protected SEXP object.  So a call to this function would make the previous object in that slot become unprotected" behavior.

Reply:
got it, thank you for sharing.